From 58f413c1fa4e1befb6b40dab4258f4e78baed160 Mon Sep 17 00:00:00 2001
From: John Zhuge <jzhuge@cloudera.com>
Date: Sat, 7 Oct 2017 00:16:56 -0700
Subject: [PATCH 2654/2848] HDFS-12494. libhdfs SIGSEGV in
 setTLSExceptionStrings. Contributed by John
 Zhuge.

(cherry picked from commit 2856eb207bfb206f22a6266f42cad0257083ab94)

Change-Id: I2d11c1eace0d0e7607d698822e498a810a26a385
---
 .../src/main/native/libhdfs/jni_helper.c           |   12 +++++++-----
 1 file changed, 7 insertions(+), 5 deletions(-)

diff --git a/hadoop-hdfs-project/hadoop-hdfs/src/main/native/libhdfs/jni_helper.c b/hadoop-hdfs-project/hadoop-hdfs/src/main/native/libhdfs/jni_helper.c
index e7c08aa..c45d598 100644
--- a/hadoop-hdfs-project/hadoop-hdfs/src/main/native/libhdfs/jni_helper.c
+++ b/hadoop-hdfs-project/hadoop-hdfs/src/main/native/libhdfs/jni_helper.c
@@ -522,19 +522,21 @@ JNIEnv* getJNIEnv(void)
     /* Create a ThreadLocalState for this thread */
     state = threadLocalStorageCreate();
     if (!state) {
+      mutexUnlock(&jvmMutex);
       fprintf(stderr, "getJNIEnv: Unable to create ThreadLocalState\n");
       return NULL;
     }
-    state->env = getGlobalJNIEnv();
-    mutexUnlock(&jvmMutex);
-    if (!state->env) {
-      goto fail;
-    }
     if (threadLocalStorageSet(state)) {
+      mutexUnlock(&jvmMutex);
       goto fail;
     }
     THREAD_LOCAL_STORAGE_SET_QUICK(state);
 
+    state->env = getGlobalJNIEnv();
+    mutexUnlock(&jvmMutex);
+    if (!state->env) {
+      goto fail;
+    }
     return state->env;
 
 fail:
-- 
1.7.9.5

